### PR TITLE
perf: optimizing performance of router match with lrucache

### PR DESCRIPTION
**Pull request by tzssangglass**
### Description

Fixes # (issue)

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: apisix/core/improve.lua [hunk 1]
```diff
@@ -0,0 +1,110 @@
+--
```
Comment (membphis) [inline apisix/core/improve.lua:1]: we should use `ai.lua`

**Review by membphis — CHANGES_REQUESTED** [review]

File: apisix/core/improve.lua [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+function  _M.routes_analyze(routes)
+    local vars_flag = false
+    local filter_flag = false
+    local prefix_match_flag = false
+
+    for _, route in ipairs(routes) do
+        if type(route) == "table" then
+            if route.vars then
+                vars_flag = true
+                break
```
Comment (membphis) [inline apisix/core/improve.lua:62]: we should remove `break` here

File: apisix/core/improve.lua [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local require         = require
+local core            = require("apisix.core")
+local ipairs          = ipairs
+local type            = type
+
+local route_lrucache = core.lrucache.new({
+    count = 512
```
Comment (membphis) [inline apisix/core/improve.lua:23]: we need to create a new route_lrucache object when enable the `route cache` logic

File: apisix/core/improve.lua [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+local function match_route(ctx)
+    orig_router_match(ctx)
+    -- replace the router match
+    return ctx.matched_route or false
+end
+
+
+local function ai_match(ctx)
+    local route_ckey = ctx.var.uri .. "-" .. ctx.var.method .. "-" ..
```
Comment (membphis) [inline apisix/core/improve.lua:41]: todo: generate the cache key dynamic

File: apisix/core/improve.lua (hunk 2 above)
Comment (membphis) [reply apisix/core/improve.lua:62]: new code style:

```lua
local route_flags = {}

for route in ... do
    for k, _ pairs(route) do
        route_flags[k] = true
    end
end

if route_flags.vars or route_flags.filter_funs then
    ...
else
    ...
end
```

File: apisix/core/improve.lua [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        enable_route_cache = false
+        router.match = orig_router_match
+    else
+        enable_route_cache = true
+        router.match = ai_match
+    end
+end
+
+
+function _M.enable_route_cache()
```
Comment (membphis) [inline apisix/core/improve.lua:96]: we can remove this `function`

**Comment by membphis** [comment]
and we should provide more information about the new performance before we merge this PR

File: apisix/init.lua [hunk 6]
```diff
@@ -374,7 +375,9 @@ function _M.http_access_phase()
     api_ctx.var.real_request_uri = api_ctx.var.request_uri
     api_ctx.var.request_uri = api_ctx.var.uri .. api_ctx.var.is_args .. (api_ctx.var.args or "")
 
+    improve.router_match(api_ctx)
     router.router_http.match(api_ctx)
+    improve.router_match_post(api_ctx)
```
[the remark this answers is not in the document]
Comment (tzssangglass) [reply apisix/init.lua]: done

File: apisix/core/improve.lua (hunk 2 above)
Comment (tzssangglass) [reply apisix/core/improve.lua]: done

File: apisix/core/improve.lua (hunk 5 above)
Comment (tzssangglass) [reply apisix/core/improve.lua]: done

File: apisix/core/improve.lua (hunk 4 above)
Comment (tzssangglass) [reply apisix/core/improve.lua]: done

File: apisix/core/improve.lua (hunk 3 above)
Comment (tzssangglass) [reply apisix/core/improve.lua]: add TODO

File: apisix/http/router/radixtree_uri.lua [hunk 7]
```diff
@@ -18,6 +18,7 @@ local require = require
 local core = require("apisix.core")
 local base_router = require("apisix.http.route")
 local get_services = require("apisix.http.service").services
+local routes_analyze = require("apisix.core.improve").routes_analyze
```
[the remark this answers is not in the document]
Comment (tzssangglass) [reply apisix/http/router/radixtree_uri.lua]: done

**Comment by tzssangglass** [comment]
The following tests were done on my development machine.

The number of workers is 1, disable access log, and disable the proxy-mirror and proxy-cache plugins

### master branch:

```
Requests/sec:   9883.91
Transfer/sec:      1.81MB

Requests/sec:  10452.79
Transfer/sec:      1.91MB

Requests/sec:  10500.38
Transfer/sec:      1.92MB
```

flamegraph: master


### perf

```
Requests/sec:  11641.22
Transfer/sec:      2.13MB

Requests/sec:  11447.80
Transfer/sec:      2.10MB

Requests/sec:  11415.72
Transfer/sec:      2.09MB
```

flamegraph: 
perf

File: apisix/core/ai_plane.lua [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+function  _M.routes_analyze(routes)
+    -- TODO: we need to add a option in config.yaml to enable this feature(default is true)
+    local route_flags = core.table.new(0, 2)
+    for _, route in ipairs(routes) do
+        if route.vars then
+            route_flags.vars = true
+        end
+
+        if route.filter_fun then
```
Comment (membphis) [inline apisix/core/ai_plane.lua:60]: need to check `route.remote_addr` and `route.remote_addrs`

Comment (tzssangglass) [reply apisix/core/ai_plane.lua:60]: why? ctx.var.remote_addr is part of cache key.

File: apisix/core/ai_plane.lua [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+local function match_route(ctx)
+    orig_router_match(ctx)
+    return ctx.matched_route or false
+end
+
+
+local function ai_match(ctx)
+    -- TODO: we need to generate cache key dynamically
+    local key = ctx.var.uri .. "-" .. ctx.var.method .. "-" ..
+                ctx.var.host .. "-" .. ctx.var.remote_addr
```
Comment (spacewander) [inline apisix/core/ai_plane.lua:41]: Is it suitable to cache remote_addr? The range can be very hugh.

File: apisix/core/ai_plane.lua [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    end
+end
+
+
+function  _M.routes_analyze(routes)
+    -- TODO: we need to add a option in config.yaml to enable this feature(default is true)
+    local route_flags = core.table.new(0, 2)
+    for _, route in ipairs(routes) do
+        if route.vars then
+            route_flags.vars = true
```
Comment (spacewander) [inline apisix/core/ai_plane.lua:57]: We could simplify it with an `enable_ai_plane` flag?
```suggestion
            enable_ai_plane = false
```

**Review by membphis — APPROVED** [review]

File: apisix/core/ai_plane.lua [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        ctx.matched_route = route_cache
+    end
+end
+
+
+function  _M.routes_analyze(routes)
+    -- TODO: we need to add a option in config.yaml to enable this feature(default is true)
+    local route_flags = core.table.new(0, 2)
+    for _, route in ipairs(routes) do
+        if route.vars then
```
Comment (membphis) [inline apisix/core/ai_plane.lua:56]:
```lua
for k, _ in pairs(route) do
  route_flags[k] = true
end
```

much simple

File: apisix/core/ai_plane.lua (hunk 10 above)
Comment (membphis) [reply apisix/core/ai_plane.lua:57]: in this case, `enable_ai_plane` is not a good name

File: apisix/core/ai_plane.lua (hunk 9 above)
Comment (tzssangglass) [reply apisix/core/ai_plane.lua:41]: so as long as the user choose `remote_addr` in route config, disable route cache?

File: apisix/core/ai_plane.lua (hunk 11 above)
Comment (tzssangglass) [reply apisix/core/ai_plane.lua:55]: done

File: apisix/core/ai_plane.lua (hunk 9 above)
Comment (tzssangglass) [reply apisix/core/ai_plane.lua]: done, remove `remote_addr` in cache key

File: apisix/core/ai_plane.lua (hunk 8 above)
Comment (tzssangglass) [reply apisix/core/ai_plane.lua:59]: add

File: apisix/core/ai.lua [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    -- if the version has not changed, use the cached route
+    if route_cache then
+        ctx.matched_route = route_cache
+    end
+end
+
+
+function  _M.routes_analyze(routes)
+    -- TODO: we need to add a option in config.yaml to enable this feature(default is true)
+    local route_flags = core.table.new(0, 2)
```
Comment (soulbird) [inline apisix/core/ai.lua:53]: It seems that `route_flags` is a `bool` type enough, no need for a `table`?

Comment (tzssangglass) [reply apisix/core/ai.lua:53]: In fact, the way it is written here is to prepare for subsequent optimizations, and we need to turn on different optimizations depending on different combinations of flags.

Comment (soulbird) [reply apisix/core/ai.lua:53]: ok~

**Review by soulbird — APPROVED** [review]

**Review by spacewander — APPROVED** [review]

**Pull request merged**